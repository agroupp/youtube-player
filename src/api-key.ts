export const YOUTUBE_API_KEY = 'AIzaSyCLWVzVwKlEmVzdumw4qWqkeEEJgRZq3LA';