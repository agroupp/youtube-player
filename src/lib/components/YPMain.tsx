import * as React from 'react';

import { YPPlaylistComponent } from './YPPlaylist';
import { YPPlayerComponent } from './YPPlayer';
import { Video } from '../interfaces/video';

interface State {
    active: string;
    videos: Array<Video>;
}

interface Props {
    apiKey: string;
}

export class YPMainComponent extends React.Component<Props, State> {
    state: State = { active: '', videos: []};

    constructor(props: Props) {
        super(props);
        this.onVideoAdd = this.onVideoAdd.bind(this);
        this.onPlayerFinished = this.onPlayerFinished.bind(this);
        this.onVideoRemove = this.onVideoRemove.bind(this);
    }

    /**
     * Video Add handler
     * @param video 
     */
    onVideoAdd(video: Video) {
        if (this.state.videos.filter(v => v.id === video.id).length === 0) {
            let videos: Array<Video> = this.state.videos;
            videos.push(video);
            this.setState({videos: videos});
            if (this.state.videos.length === 1) {
                this.setState({ active: video.id });
            }
        }
    }

    /**
     * Player finished handler
     */
    onPlayerFinished() {
        let videos: Array<Video> = this.state.videos.filter(v => v.id !== this.state.active);
        if (videos.length > 0) {
            this.setState({videos: videos, active: videos[0].id});
        } else {
            this.setState({videos: [], active: ''});
        }
    }

    /**
     * Video remove from list handler
     * @param id Youtube video ID
     */
    onVideoRemove(id: string) {
        if (this.state.active === id) {
            this.onPlayerFinished();
        } else {
            let videos: Array<Video> = this.state.videos.filter(v => v.id !== id);
            this.setState({videos: videos});
        }
    }
    
    render() {
        return (
            <div className="container">
                <div className="row">
                    <div className="col-12 col-lg-4">
                        <YPPlaylistComponent
                            apiKey={this.props.apiKey}
                            onVideoAdd={this.onVideoAdd}
                            onVideoRemove={this.onVideoRemove}
                            videos={this.state.videos}
                            active={this.state.active}
                        />
                    </div>
                    <div className="col-12 col-lg-7">
                        <YPPlayerComponent 
                            videoId={this.state.active}
                            onPlayerFinished={this.onPlayerFinished}
                        />
                    </div>
                </div>
            </div>
        );
    }
}