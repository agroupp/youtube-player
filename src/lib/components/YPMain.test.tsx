import * as React from 'react';
import * as ReactDOM from 'react-dom';

/* It works only this way because there is no renderer in @types */
const renderer = require('react-test-renderer');

import { YPMainComponent } from './YPMain';
import { YOUTUBE_API_KEY } from '../../api-key';

describe('<YPMainComponent />', function(){
  it('should render without crashing by React', () => {
    const div = document.createElement('div');
    ReactDOM.render(<YPMainComponent apiKey={YOUTUBE_API_KEY} />, div);
  });

  it('should render and create JSON tree', () => {
    const component = renderer.create(<YPMainComponent apiKey={YOUTUBE_API_KEY} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
