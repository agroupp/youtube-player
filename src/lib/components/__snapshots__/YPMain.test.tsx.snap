// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<YPMainComponent /> should render and create JSON tree 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-12 col-lg-4"
    >
      <div
        className="yp-playlist-wrapper card h-100"
      >
        <div
          className="card-body h-100 d-flex flex-column"
        >
          <div
            className="row"
          >
            <div
              className="yp-input-wrapper col-12"
            >
              <form
                className="form-inline"
                onSubmit={[Function]}
              >
                <div
                  className="form-group col-9 w-100 px-0 pr-1 m-0"
                >
                  <input
                    className="form-control w-100"
                    onChange={[Function]}
                    placeholder="Enter Video URL or Video ID"
                    style={
                      Object {
                        "borderColor": "rgba(0,0,0,0.125)",
                      }
                    }
                    type="text"
                    value=""
                  />
                </div>
                <button
                  className="btn btn-primary col-3"
                  disabled={true}
                  type="submit"
                >
                  Add
                </button>
              </form>
              <div
                className="row yp-invalid-feedback-row"
              >
                <div
                  className="invalid-feedback col-12"
                  style={
                    Object {
                      "display": "none",
                    }
                  }
                >
                  Wrong Video Id or URL
                </div>
                <div
                  className="col-12 text-success"
                  style={
                    Object {
                      "display": "none",
                    }
                  }
                >
                  <span>
                    Processing...
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            className="yp-list-row"
          >
            <div
              className="yp-list-wrapper"
            >
              <ul
                className="list-group"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="col-12 col-lg-7"
    >
      <div
        className="card"
      >
        <div
          className="card-body h-100 d-flex flex-column"
        >
          <div
            className="embed-responsive embed-responsive-16by9"
            id="player"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
